gpflow/kullback_leiblers.py

```diff
@@ -71,32 +73,45 @@ def gauss_kl(q_mu, q_sqrt, K=None, *, K_cholesky=None):
     is_white = (K is None) and (K_cholesky is None)
     is_diag = len(q_sqrt.shape) == 2
 
-    M, B = tf.shape(q_mu)[0], tf.shape(q_mu)[1]
+    shape_constraints = [
+        (q_mu, ["M", "L"]),
+        (q_sqrt, (["M", "L"] if is_diag else ["L", "M", "M"])),
+    ]
+    if not is_white:
+        if K is not None:
+            shape_constraints.append((K, (["L", "M", "M"] if len(K.shape) == 3 else ["M", "M"])))
+        else:
+            shape_constraints.append(
+                (K_cholesky, (["L", "M", "M"] if len(K_cholesky.shape) == 3 else ["M", "M"]))
+            )
```

Comment:
just to give you some idea of how nuts idris is, you could write this all as part of the type and it would check it at compile time. yes, including the function calls.

Reply:
I'd love that...